src/nupic/research/TP_shim.py

```diff
@@ -105,6 +105,21 @@ def compute(self, bottomUpInput, enableLearn, computeInfOutput=None):
     return output
 
 
+  def topDownCompute(self, topDownIn=None):
```

Comment:
Could you create a mixin class that defines functions like this and add it as an additional base class for each shim? Just trying to think of a way to avoid so much duplicated code.